src/object.d

```diff
@@ -4881,3 +4881,90 @@ void __ArrayDtor(T)(T[] a)
     foreach_reverse (ref T e; a)
         e.__xdtor();
 }
+
+/***********************************************************
+ * The compiler lowers expressions of `cast(TTo[])TFrom[]` to
+ * this implementation.
+ * Params:
+ *  from = the array to reinterpret-cast
+ * Returns:
+ *  `from` reinterpreted as `TTo[]`
+ */
+TTo[] __ArrayCast(TFrom, TTo)(TFrom[] from) @trusted
+{
+    const nbytes = from.length * TFrom.sizeof;
+
+    struct Array
+    {
+        size_t length;
+        void* ptr;
+    }
+
+    if ((nbytes % TTo.sizeof) != 0)
+    {
+        import core.internal.string : unsignedToTempString;
+        import core.stdc.stdlib : alloca;
+
+        const(char)[][8] msgComponents =
+        [
+            "Cannot cast `"
+            , TFrom.stringof
+            , "` to `"
+            , TTo.stringof
+            , "`, an array of size "
+            , unsignedToTempString(nbytes)
+            , " does not align on an array of size "
+            , unsignedToTempString((nbytes / TTo.sizeof) + (nbytes % TTo.sizeof))
+        ];
+
+        // convert discontiguous `msgComponents` to contiguous string on the stack
+        char[] msg;
+
+        auto array = cast(Array*)&msg;
+
+        foreach(m ; msgComponents)
+            array.length += m.length;
+        array.ptr = alloca(array.length);
+
+        size_t index = 0;
+        foreach(m ; msgComponents)
+            foreach(c; m)
+                msg[index++] = c;
+
+        version (D_Exceptions)
+        {
+            import core.exception : throwStaticError;
+            throwStaticError(cast(string)msg);
+        }
+        else
+        {
+            assert(false, cast(string)msg);
```

Comment:
I don't know how `assert(false)` behaves but `assert(0)` will halt the program when assertions are turned off.

Reply:
`assert(false` will do the same because it's the same: https://run.dlang.io/is/AVSUFK

@JinShil we typically write `assert(0, ...`

Reply:
>  we typically write assert(0, ...

IMO we shot stop doing that.  It's a condition, not a number.

Reply:
@jacob-carlborg I tested this out and you're right.  If the first argument is `0` or `false` it still runs when asserts are disabled.  However, I can find no mention of this behavior in the spec.  If this is by design, it'd great if you could document it, and then I could potentially remove throwing the static `Error` altogether.  

Kindly asking for advise from all reviewers on how to proceed.

Reply:
Found this in [the spec](https://dlang.org/spec/expression.html#assert_expressions).

> The implementation may handle the case of the first AssignExpression evaluating at compile time to false differently in that in release mode it may simply generate a HLT instruction or equivalent. 

I guess I'l just remove the static `Error` and rely on `assert(false)` for all implementations, then.  It should keep the functions `@trusted` even in release mode.

Reply:
> Please don't.

I'd be happy to revert back to throwing a static `Error`, but I need consensus from reviewers or decree from @WalterBright

Reply:
I know there's been some disagreement about the behavior of `assert(0/false)` because not all CPUs have the `HLT` instruction, but perhaps that covered with the "or equivalent" in the spec.